tsdb/query_executor.go

```diff
@@ -0,0 +1,954 @@
+package tsdb
+
+import (
+	"errors"
+	"fmt"
+	"log"
+	"os"
+	"runtime"
+	"sort"
+	"strings"
+
+	"github.com/influxdb/influxdb/influxql"
+	"github.com/influxdb/influxdb/meta"
+)
+
+type QueryExecutor struct {
+	// The meta store for accessing and updating cluster and schema data.
+	MetaStore interface {
+		Database(name string) (*meta.DatabaseInfo, error)
+		Databases() ([]meta.DatabaseInfo, error)
+		User(name string) (*meta.UserInfo, error)
+		AdminUserExists() (bool, error)
+		Authenticate(username, password string) (*meta.UserInfo, error)
+		RetentionPolicy(database, name string) (rpi *meta.RetentionPolicyInfo, err error)
+	}
+
+	// Executes statements relating to meta data.
+	MetaStatementExecutor interface {
+		ExecuteStatement(stmt influxql.Statement) *influxql.Result
+	}
+
+	// The stats service to report to
+	Stats interface {
+		Add(key string, delta int64)
+		Inc(key string)
+		Name() string
+		Walk(f func(string, int64))
+	}
+
+	Logger *log.Logger
+
+	// the local daata store
+	store *Store
+}
+
+func NewQueryExecutor(store *Store) *QueryExecutor {
+	return &QueryExecutor{
+		store:  store,
+		Logger: log.New(os.Stderr, "[query] ", log.LstdFlags),
+	}
+}
+
+// Begin is for influxql/engine.go to use to get a transaction object to start the query
+func (q *QueryExecutor) Begin() (influxql.Tx, error) {
+	return newTx(q.MetaStore, q.store), nil
+}
+
+// Authorize user u to execute query q on database.
+// database can be "" for queries that do not require a database.
+// If u is nil, this means authorization is disabled.
```

Comment:
This comment makes me think if I pass in a nil user, it will succeed, but it actually will fail.